File: src/lichess/player.js
import { Chess } from "chess.js";
import { encrypt } from "../steg/crypto.js";
import { rankCandidates } from "../steg/stockfish.js";
import { createAIGame, makeMove, streamGame, challengeUser, streamAccountEvents, resignGame } from "./api.js";

function bytesToBits(bytes) {
  return Array.from(bytes)
    .map((b) => b.toString(2).padStart(8, "0"))
    .join("");
}

function capacityBits(numMoves) {
  return Math.max(0, Math.floor(Math.log2(numMoves)));
}

const ENCODING_BITS = 4;
const ENCODING_PATTERNS = 1 << ENCODING_BITS; // 16

// Deterministic quality sort — used by both encoder and decoder.
// Better moves get lower indices so the encoding prefers them statistically.
const PIECE_VAL = { p: 1, n: 3, b: 3, r: 5, q: 9 };
function moveScore(m) {
  let s = 0;
  if (m.captured) s -= PIECE_VAL[m.captured] * 100;  // captures LAST — quiet moves encode first
  if (m.promotion) s += PIECE_VAL[m.promotion] * 50; // promotions near front
  // Penalise rook/queen shuffles along the back rank (they keep appearing at index 0)
  if ((m.piece === "r" || m.piece === "q") && m.from[1] === m.to[1] && m.from[1] === "1") s -= 40;
  // Deprioritize non-castling king moves; castling flags are "k" (kingside) and "q" (queenside)
  if (m.piece === "k" && !m.flags.includes("k") && !m.flags.includes("q")) s -= 200;
  return s;
}
function qualitySort(a, b) {
  const diff = moveScore(b) - moveScore(a); // descending score
  if (diff !== 0) return diff;
  const uciA = a.from + a.to + (a.promotion || "");
  const uciB = b.from + b.to + (b.promotion || "");
  return uciA.localeCompare(uciB); // tiebreak by UCI for determinism
}

// Stateful encoding session — works incrementally as the game progresses.
// Encodes bits into White's moves; Black (AI) plays freely.
class StegSession {
  constructor(allBits) {
    this.allBits = allBits;
    this.bitIndex = 0;
    this.chess = new Chess();
  }

  get isDone() {
    return this.bitIndex >= this.allBits.length;
  }

  get progress() {
    return this.allBits.length === 0
      ? 1
      : this.bitIndex / this.allBits.length;
  }

  // Apply any move (White or Black) — accepts SAN or UCI ("e2e4")
  applyMove(move) {
    if (/^[a-h][1-8][a-h][1-8][qrbnQRBN]?$/.test(move)) {
      const from = move.slice(0, 2);
      const to = move.slice(2, 4);
      const promotion = move.length === 5 ? move[4] : undefined;
      this.chess.move({ from, to, promotion });
    } else {
      this.chess.move(move);
    }
  }

  // Pick the next White move that encodes the next bits chunk.
  // Uses Stockfish to choose the best-looking move within the encoding equivalence class.
  // Returns { san, uci } — san for display, uci for Lichess API
  async nextEncodingMove() {
    if (this.isDone) return null;
    let verboseMoves = this.chess.moves({ verbose: true });
    if (verboseMoves.length === 0 || this.chess.isGameOver()) return null;

    // Quality sort determines the INDEX MAPPING — decoder mirrors this exactly.
    // Stockfish only picks WHICH candidate within the equivalence class to play.
    verboseMoves.sort(qualitySort);

    // Variable-capacity encoding: floor(log2(n)) bits, capped at ENCODING_BITS.
    // Encoding always progresses regardless of how many moves are available.
    // 1 move → 0 bits (forced); 2–3 → 1 bit; 4–7 → 2 bits; 8–15 → 3 bits; 16+ → 4 bits.
    const moveBits = Math.min(ENCODING_BITS, Math.floor(Math.log2(verboseMoves.length)));

    if (moveBits === 0) {
      // Only 1 legal non-king move — forced, encode 0 bits. Prefer non-capture.
      const m = verboseMoves.find((mv) => !mv.captured) ?? verboseMoves[0];
      return { san: m.san, uci: m.from + m.to + (m.promotion || "") };
    }

    const PATTERNS = 1 << moveBits; // 2, 4, 8, or 16

    const remaining = this.allBits.length - this.bitIndex;
    const bitsToUse = Math.min(moveBits, remaining);
    const b = parseInt(
      this.allBits.slice(this.bitIndex, this.bitIndex + bitsToUse).padEnd(moveBits, "0"),
      2
    );

    // Collect all valid candidates: indices b, b+PATTERNS, b+2*PATTERNS, …
    // (skip any that accidentally deliver checkmate mid-encoding)
    let candidates = [];
    for (let idx = b; idx < verboseMoves.length; idx += PATTERNS) {
      const m = verboseMoves[idx];
      const temp = new Chess(this.chess.fen());
      temp.move(m.san);
      if (!temp.isCheckmate()) candidates.push(m);
    }
    if (candidates.length === 0) candidates.push(verboseMoves[b]); // extremely rare fallback

    // Prefer non-captures — captures create material imbalance and cause early resignation.
    // The decoder accepts any candidate at index b, b+PATTERNS, … (all decode to the same bits).
    const nonCaptures = candidates.filter((m) => !m.captured);
    if (nonCaptures.length > 0) candidates = nonCaptures;

    // Use Stockfish to pick the highest-quality candidate
    let chosen;
    if (candidates.length === 1) {
      chosen = candidates[0];
    } else {
      const candidateUCIs = candidates.map((m) => m.from + m.to + (m.promotion || ""));
      const ranked = await rankCandidates(this.chess.fen(), candidateUCIs);
      chosen = candidates.find((m) => m.from + m.to + (m.promotion || "") === ranked[0]) ?? candidates[0];
    }

    this.bitIndex += bitsToUse;
    return { san: chosen.san, uci: chosen.from + chosen.to + (chosen.promotion || "") };
  }
}

// Build bits payload from plaintext + password
async function buildBits(plaintext, password) {
  const cipherB64 = await encrypt(plaintext, password);
  const cipherBytes = Uint8Array.from(atob(cipherB64), (c) => c.charCodeAt(0));
  const lenBits = cipherBytes.length.toString(2).padStart(8, "0");
  return lenBits + bytesToBits(cipherBytes);
}

// Play the full encoded game on Lichess vs a human or AI (level 1 fallback).
// Calls onEvent({ type, ... }) for UI updates.
export async function playEncodedGame(plaintext, password, opponentUsername, onEvent = () => {}) {
  const allBits = await buildBits(plaintext, password);
  const session = new StegSession(allBits);

  let gameId;
  if (opponentUsername) {
    onEvent({ type: "challenging", opponent: opponentUsername });
    const result = await challengeUser(opponentUsername);
    gameId = (result.challenge ?? result).id;
    onEvent({ type: "waiting", gameId, opponent: opponentUsername, url: `https://lichess.org/${gameId}` });

    const acceptResult = await Promise.race([
      (async () => {
        for await (const event of streamAccountEvents()) {
          if (event.type === "gameStart" && event.game?.gameId === gameId) return "accepted";
          if (event.type === "challengeDeclined") return "declined";
        }
        return "declined";
      })(),
      new Promise((resolve) => setTimeout(() => resolve("timeout"), 2 * 60 * 1000)),
    ]);
    if (acceptResult === "timeout") {
      onEvent({ type: "timeout" });
      return;
    }
    if (acceptResult === "declined") {
      onEvent({ type: "declined", opponent: opponentUsername });
      return;
    }
    onEvent({ type: "started", gameId, url: `https://lichess.org/${gameId}` });
  } else {
    // AI fallback for solo testing
    onEvent({ type: "creating" });
    const game = await createAIGame("white", 1);
    gameId = game.id;
    onEvent({ type: "created", gameId, url: `https://lichess.org/${gameId}` });
  }

  const whiteMoves = [];

  for await (const event of streamGame(gameId)) {
    if (event.type !== "gameFull" && event.type !== "gameState") continue;

    const state = event.type === "gameFull" ? event.state : event;
    const serverMoves = state.moves ? state.moves.split(" ").filter(Boolean) : [];

    // Always rebuild board from server's authoritative move list
    const rebuilt = new Chess();
    for (const uci of serverMoves) {
      rebuilt.move({ from: uci.slice(0, 2), to: uci.slice(2, 4), promotion: uci[4] || undefined });
    }
    session.chess = rebuilt;

    if (serverMoves.length > 0) {
      onEvent({ type: "fen", fen: rebuilt.fen() });
    }

    const isWhiteTurn = rebuilt.turn() === "w";

    if (isWhiteTurn && !rebuilt.isGameOver()) {
      let move;
      if (!session.isDone) {
        // Still encoding — pick a steganographic move (Stockfish-ranked)
        move = await session.nextEncodingMove();
        if (!move) break;
        if (session.isDone) {
          // Just encoded the last bit — notify UI
          onEvent({ type: "encoded", gameId, url: `https://lichess.org/${gameId}`, whiteMoves: [...whiteMoves] });
        }
      } else {
        // Encoding done — resign immediately so the game is archived and decodable
        try { await resignGame(gameId); } catch { /* ignore — game may already be over */ }
        onEvent({ type: "done", status: "resign", gameId, url: `https://lichess.org/${gameId}`, whiteMoves });
        break;
      }

      await makeMove(gameId, move.uci);
      whiteMoves.push(move.san);
      rebuilt.move({ from: move.uci.slice(0, 2), to: move.uci.slice(2, 4), promotion: move.uci[4] || undefined });

      onEvent({
        type: "move",
        move: move.san,
        fen: rebuilt.fen(),
        moveNum: whiteMoves.length,
        progress: Math.min(session.progress, 1),
        gameId,
      });
    } else if (!isWhiteTurn) {
      onEvent({ type: "ai_thinking" });
    }

    // Break only when game ends naturally
    if (state.status && !["created", "started"].includes(state.status)) {
      if (!session.isDone) {
        onEvent({ type: "incomplete", gameId, url: `https://lichess.org/${gameId}`, whiteMoves });
      } else {
        onEvent({ type: "done", status: state.status, gameId, url: `https://lichess.org/${gameId}`, whiteMoves });
      }
      break;
    }
  }

  return { gameId, url: `https://lichess.org/${gameId}`, whiteMoves };
}

File: lichess/player.py
import asyncio
import chess
import chess.engine

from steg.crypto import encrypt
from steg.quality import (
    capacity_bits,
    get_sorted_moves,
    ENCODING_BITS,
    PREAMBLE_MOVES,
    PREAMBLE_LENGTH,
)
from lichess.api import (
    create_ai_game,
    make_move,
    stream_game,
    resign_game,
    challenge_user,
    stream_account_events,
)

STOCKFISH_PATH = "stockfish"


def _bytes_to_bits(data: bytes) -> str:
    return "".join(f"{b:08b}" for b in data)


async def _open_engine():
    try:
        _, engine = await chess.engine.popen_uci(STOCKFISH_PATH)
        return engine
    except Exception:
        return None  # Stockfish not installed — skip quality ranking


async def _rank_candidates(
    engine, board: chess.Board, candidates: list[chess.Move]
) -> list[chess.Move]:
    if not engine or len(candidates) <= 1:
        return candidates
    try:
        multipv = min(20, sum(1 for _ in board.legal_moves))
        info_list = await asyncio.wait_for(
            engine.analyse(board, chess.engine.Limit(time=0.1), multipv=multipv),
            timeout=2.0,
        )
        if isinstance(info_list, dict):
            info_list = [info_list]
        scored: dict[chess.Move, int] = {}
        for info in info_list:
            pv = info.get("pv", [])
            if pv:
                scored[pv[0]] = info["score"].relative.score(mate_score=100_000) or 0
        return sorted(candidates, key=lambda m: -(scored.get(m, -99_999)))
    except Exception:
        return candidates


# ─── StegSession ─────────────────────────────────────────────────────────────

class StegSession:
    def __init__(self, all_bits: str):
        self.all_bits = all_bits
        self.bit_index = 0
        self.board = chess.Board()
        self._white_move_count = 0

    @property
    def preamble_done(self) -> bool:
        return self._white_move_count >= PREAMBLE_LENGTH

    @property
    def is_done(self) -> bool:
        return self.preamble_done and self.bit_index >= len(self.all_bits)

    @property
    def progress(self) -> float:
        if not self.preamble_done:
            return 0.0
        if not self.all_bits:
            return 1.0
        return self.bit_index / len(self.all_bits)

    def record_white_move(self):
        self._white_move_count += 1

    async def next_encoding_move(self, engine) -> dict | None:
        # Play fixed preamble moves before encoding begins
        if not self.preamble_done:
            uci = PREAMBLE_MOVES[self._white_move_count]
            move = chess.Move.from_uci(uci)
            return {"san": self.board.san(move), "uci": uci}

        if self.is_done:
            return None

        moves = get_sorted_moves(self.board)
        if not moves or self.board.is_game_over():
            return None

        move_bits = capacity_bits(len(moves))

        if move_bits == 0:
            # Only one legal move — forced, encode 0 bits
            m = next((mv for mv in moves if not self.board.is_capture(mv)), moves[0])
            return {"san": self.board.san(m), "uci": m.uci()}

        patterns = 1 << move_bits
        remaining = len(self.all_bits) - self.bit_index
        bits_to_use = min(move_bits, remaining)
        b = int(
            self.all_bits[self.bit_index : self.bit_index + bits_to_use].ljust(
                move_bits, "0"
            ),
            2,
        )

        # All indices b, b+patterns, b+2*patterns … decode to the same bits
        candidates = []
        for idx in range(b, len(moves), patterns):
            m = moves[idx]
            temp = self.board.copy()
            temp.push(m)
            if not temp.is_checkmate():
                candidates.append(m)
        if not candidates:
            candidates = [moves[b]]

        # Prefer non-captures — avoids material swings that cause early resignation
        non_captures = [m for m in candidates if not self.board.is_capture(m)]
        if non_captures:
            candidates = non_captures

        # Let Stockfish pick the highest-quality candidate
        candidates = await _rank_candidates(engine, self.board, candidates)

        chosen = candidates[0]
        self.bit_index += bits_to_use
        return {"san": self.board.san(chosen), "uci": chosen.uci()}


# ─── Main entry point (async generator — yields events for the UI) ────────────

async def play_encoded_game(plaintext: str, password: str, opponent_username: str | None = None):
    cipher_bytes = encrypt(plaintext, password)
    all_bits = f"{len(cipher_bytes):08b}" + _bytes_to_bits(cipher_bytes)

    session = StegSession(all_bits)
    engine = await _open_engine()

    try:
        game_id: str | None = None

        if opponent_username:
            yield {"type": "challenging", "opponent": opponent_username}
            result = await challenge_user(opponent_username)
            game_id = (result.get("challenge") or result)["id"]
            yield {
                "type": "waiting",
                "gameId": game_id,
                "opponent": opponent_username,
                "url": f"https://lichess.org/{game_id}",
            }

            accept_result = "declined"
            try:
                async with asyncio.timeout(120):
                    async for event in stream_account_events():
                        if (
                            event.get("type") == "gameStart"
                            and event.get("game", {}).get("gameId") == game_id
                        ):
                            accept_result = "accepted"
                            break
                        if event.get("type") == "challengeDeclined":
                            break
            except TimeoutError:
                yield {"type": "timeout"}
                return

            if accept_result != "accepted":
                yield {"type": "declined", "opponent": opponent_username}
                return

            yield {"type": "started", "gameId": game_id, "url": f"https://lichess.org/{game_id}"}

        else:
            yield {"type": "creating"}
            game = await create_ai_game("white", 1)
            game_id = game["id"]
            yield {"type": "created", "gameId": game_id, "url": f"https://lichess.org/{game_id}"}

        white_moves: list[str] = []

        async for event in stream_game(game_id):
            if event["type"] not in ("gameFull", "gameState"):
                continue

            state = event.get("state", event) if event["type"] == "gameFull" else event
            server_moves = [m for m in state.get("moves", "").split() if m]

            # Rebuild board from server's authoritative move list
            rebuilt = chess.Board()
            for uci in server_moves:
                rebuilt.push(chess.Move.from_uci(uci))
            session.board = rebuilt

            if server_moves:
                yield {"type": "fen", "fen": rebuilt.fen()}

            is_white_turn = rebuilt.turn == chess.WHITE

            if is_white_turn and not rebuilt.is_game_over():
                if not session.is_done:
                    move = await session.next_encoding_move(engine)
                    if not move:
                        break

                    await make_move(game_id, move["uci"])
                    white_moves.append(move["san"])
                    session.record_white_move()
                    rebuilt.push(chess.Move.from_uci(move["uci"]))

                    yield {
                        "type": "move",
                        "move": move["san"],
                        "fen": rebuilt.fen(),
                        "moveNum": len(white_moves),
                        "progress": min(session.progress, 1.0),
                        "gameId": game_id,
                    }

                    if session.is_done:
                        yield {
                            "type": "encoded",
                            "gameId": game_id,
                            "url": f"https://lichess.org/{game_id}",
                            "whiteMoves": list(white_moves),
                        }
                else:
                    try:
                        await resign_game(game_id)
                    except Exception:
                        pass
                    yield {
                        "type": "done",
                        "status": "resign",
                        "gameId": game_id,
                        "url": f"https://lichess.org/{game_id}",
                        "whiteMoves": white_moves,
                    }
                    break

            elif not is_white_turn:
                yield {"type": "ai_thinking"}

            status = state.get("status")
            if status and status not in ("created", "started"):
                if not session.is_done:
                    yield {
                        "type": "incomplete",
                        "gameId": game_id,
                        "url": f"https://lichess.org/{game_id}",
                        "whiteMoves": white_moves,
                    }
                else:
                    yield {
                        "type": "done",
                        "status": status,
                        "gameId": game_id,
                        "url": f"https://lichess.org/{game_id}",
                        "whiteMoves": white_moves,
                    }
                break

    finally:
        if engine:
            try:
                await engine.quit()
            except Exception:
                pass
